File: exercises/random/rateLimiter.ts
// TS (Class-based) Version
class RateLimiterTs {
  private requestLimit: number;
  private timeWindow: number;
  private requestLogs: Map<string, number[]>;

  constructor(requestLimit: number, timeWindow: number) {
    this.requestLimit = requestLimit;
    this.timeWindow = timeWindow;
    this.requestLogs = new Map();
  }

  isAllow(clientId: string): boolean {
    const currentTime = Date.now();
    if (!this.requestLogs.has(clientId)) {
      this.requestLogs.set(clientId, []);
    }
    const requestTimestamps = this.requestLogs.get(clientId)!;

    while (
      requestTimestamps.length > 0 &&
      requestTimestamps[0] < currentTime - this.timeWindow
    ) {
      requestTimestamps.shift();
    }

    if (requestTimestamps.length >= this.requestLimit) {
      return false;
    }

    requestTimestamps.push(currentTime);
    return true;
  }
}

// TS (Functional programming) Version
type RateLimiterFn = {
  isAllow: (clientId: string) => boolean;
};

const createRateLimiter = (
  requestLimit: number,
  timeWindow: number
): RateLimiterFn => {
  const requestLogs = new Map<string, number[]>();

  return {
    isAllow(clientId: string): boolean {
      const currentTime = Date.now();
      if (!requestLogs.has(clientId)) {
        requestLogs.set(clientId, []);
      }
      const requestTimestamps = requestLogs.get(clientId)!;

      while (
        requestTimestamps.length > 0 &&
        requestTimestamps[0] < currentTime - timeWindow
      ) {
        requestTimestamps.shift();
      }

      if (requestTimestamps.length >= requestLimit) {
        return false;
      }

      requestTimestamps.push(currentTime);
      return true;
    },
  };
};

File: exercises/random/rateLimiter.cs
// C# Version
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;

public class RateLimiter
{
    private readonly int _requestLimit;
    private readonly TimeSpan _timeWindow;
    private readonly ConcurrentDictionary<string, List<long>> _requestLogs;

    public RateLimiter(int requestLimit, TimeSpan timeWindow)
    {
        _requestLimit = requestLimit;
        _timeWindow = timeWindow;
        _requestLogs = new ConcurrentDictionary<string, List<long>>();
    }

    public bool IsAllow(string clientId)
    {
        long currentTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        var requestTimestamps = _requestLogs.GetOrAdd(clientId, _ => new List<long>());

        lock (requestTimestamps)
        {
            requestTimestamps.RemoveAll(timestamp => timestamp < currentTime - _timeWindow.TotalMilliseconds);
            if (requestTimestamps.Count >= _requestLimit)
            {
                return false;
            }
            requestTimestamps.Add(currentTime);
            return true;
        }
    }
}
